src/expr/datatype.cpp

```diff
@@ -1248,49 +644,41 @@ const DatatypeConstructorArg& DatatypeConstructor::operator[](std::string name)
       return *i;
     }
   }
-  IllegalArgument(name, "No such arg `%s' of constructor `%s'", name.c_str(), d_name.c_str());
+  std::string dname = getName();
+  IllegalArgument(name,
+                  "No such arg `%s' of constructor `%s'",
+                  name.c_str(),
+                  dname.c_str());
 }
 
 Expr DatatypeConstructor::getSelector(std::string name) const {
-  return (*this)[name].getSelector();
+  return (*this)[name].d_selector;
 }
 
 Type DatatypeConstructor::getArgType(unsigned index) const
 {
   PrettyCheckArgument(index < getNumArgs(), index, "index out of bounds");
-  return static_cast<SelectorType>((*this)[index].getType()).getRangeType();
+  return d_internal->getArgType(index).toType();
 }
 
 bool DatatypeConstructor::involvesExternalType() const{
-  for(const_iterator i = begin(); i != end(); ++i) {
-    if(! SelectorType((*i).getSelector().getType()).getRangeType().isDatatype()) {
-      return true;
-    }
-  }
-  return false;
+  return d_internal->involvesExternalType();
 }
 
 bool DatatypeConstructor::involvesUninterpretedType() const{
-  for(const_iterator i = begin(); i != end(); ++i) {
-    if(SelectorType((*i).getSelector().getType()).getRangeType().isSort()) {
-      return true;
-    }
-  }
-  return false;
+  return d_internal->involvesUninterpretedType();
 }
 
 DatatypeConstructorArg::DatatypeConstructorArg(std::string name, Expr selector)
-    : d_internal(nullptr),  // until the Node-level datatype API is activated
-      d_name(name),
-      d_selector(selector),
-      d_resolved(false)
+    : d_internal(nullptr)
 {
   PrettyCheckArgument(name != "", name, "cannot construct a datatype constructor arg without a name");
+  d_internal = std::make_shared<DTypeSelector>(name, Node::fromExpr(selector));
 }
 
 std::string DatatypeConstructorArg::getName() const
 {
-  string name = d_name;
+  string name = d_internal->getName();
```

Comment:
Add `ExprManagerScope ems(d_constructor);` to be on the safe side.

Reply:
There are numerous of these functions that are necessary to be called *before* resolution (when d_constructor is created).  Thus, this suggestion would lead to a segfault since `d_constructor` is NULL. 

Thus we can't just do this to be on the safe side, it's in fact less safe since it assumes that d_constructor is always non-null.

Reply:
Hmm, so we should consider also storing a pointer to the `ExprManager` in `DatatypeConstructorArg` (at least for the new API, doesn't matter that much for the `Expr` layer).

Reply:
Right, that's something I wanted to avoid for Expr since it would be even more hassle.